rfcs/010-per-document-access-control.md

```diff
@@ -0,0 +1,418 @@
+via an option passed at database creation time. If you create a
+database without that option, it works like any database in CouchDB
+today.
+
+This is how you create an access-enabled database:
+
+```
+PUT /database?access=true
+```
+
+This option can be set only at database creation time, it can’t be
+turned off and on while the database exists.
+
+An access-enabled database behaves like this:
+
+* only admin users can read or write to the database (as per 3.x
+  defaults)
+
+* admins can grant individual users and groups access to a database
+  using the database’s `_security` object. A special new role `_users`
+  can be used to say “all users defined in the `_users` database”.
+
+* documents created without an `_access` field are accessible to
+  db-admins only
+
+   * this allows existing databases to be replicated into an
+     access-enabled database, but granting access of individual docs to
+     specific users needs to be an explicit step handled by developers.
+
+* documents created with an `_access` field are only accessible by
+  admins and the user named inside `_access`.
+
+   * `_access: ["shirley"]`
+
+      * later iterations of this could allow for `["shirley"]` being
+        shorthand for `[{"read": "shirley", "write": "shirley"}]` for
+        more fine-grained access control, but that is out of scope for
+        this RFC.
+
+   * users can only create documents with their own username in
+     `_access`.
+
+   * admins can add any users to `_access`
+
+   * documents can only be owned by one user at any one point.
+
+      * in a 2.0 > X < 4.0 cluster, two different users could create
+        the same document with a different _access definition
+        concurrently and both get successful write responses back. As
+        with _users documents in conflict, if a document has a conflict
+        with separate _access entries, it becomes admin-only by
+        default. This case needs to be handled by an applications
+        _conflict handler.
+
+* document _ids are shared across all users. So only the first user who
+  creates the doc `_id: config` gets it. Applications need to ensure to
+  work around this and potentially prefix docs with the username before
+  writing/replicating them in.
+
+* _security members are allowed to write design docs, but they have to
+  have an `_access` field and those design docs with an `_access` field
+  are ignored on the server side. Db-admin ddocs get indexes built as
+  normal.
+
+   * you can’t access their views, no view indexes are built, their
+     validate_doc_update functions do not run on db inserts.
+
+   * this allows full pouchdb / satellite db replication, but avoids
+     problems with having 10000s of VDUs or 10000s of view indexes.
+
+* users can not remove themselves from `_access`, nor can they remove
+  the `_access` property. They can only `DELETE` a doc.
+
+* If an existing doc changes the user mentioned in `_access` or an admin
```

Comment:
What happens for a user that previously had access to the document? Is there a way to notify them that they have lost access?

Reply:
not as per this RFC, but we could consider a mode to `_changes`, e.g. `/_changes?access=true` that would include new rows with id/rev/revoked:true that clients could opt into.